Feature/4935 subgraph title margin config option

mathbraga opened the pull request:
## :bookmark_tabs: Summary

Adds the option to introduce top and/or bottom margins for subgraphs title.

Resolves #4935 

## :straight_ruler: Design Decisions

The approach is to increase height of subgraphs while keeping its children centered. Doing so creates space to move the
label down which creates the margin. The label only moves down by the value provided for top margin, and everything else moves by the sum of top + bottom margins.

The height of the subgraph is increased by `2 * total margin` because of how the `y axis` is calculated for nodes.
`y = node.y - height/2` means that if height is increased by twice the margin, the node will move up by the value of total margin, which is compensated in the code with `node.y + total margin`, returning the node to the center of its parent.

In the code, I chose to move the `y axis` of clusters themselves, that way everything inside of the cluster also moves, atleast that's what I understood from the code, so I went for that route for simplicity, other options would require me to also move every child node and edges inside each cluster.

This approach fails if a cluster has external connections, since it is considered a normal node with children, meaning it can't be repositioned as a cluster, from what I attempted, interacting with its `y axis` only moves its rectangle and not the whole node. I might be missing something, but from my understanding, all of its child nodes and externally connected nodes have to be repositioned along with every related edge individually. For this reason I asked in #4935 if title margins for this case could be moved to a new separate issue.

In summary, if I want to add 10 margin for top and 5 for bottom. I increase height by 30 (15x2), keep children centered, and move subgraph label down by 10 (only top margin):
![image](https://github.com/mermaid-js/mermaid/assets/31048764/3f5b8137-2a1b-4a50-80f9-49efc5f16013)

I didn't add anything to the docs since I couldn't find how exactly it works with config changes. I looked around in the code and assumed config docs are generated automatically based on types.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [ ] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

sidharthv96 commented on packages/mermaid/src/utils/getSubGraphTitleMargins.ts:
```diff
@@ -0,0 +1,17 @@
+  subGraphTitleBottomMargin: number;
+  subGraphTitleTotalMargin: number;
+} => {
+  const subGraphTitleTopMargin = getConfig().flowchart?.subGraphTitleMargin?.top || 0;
+  const subGraphTitleBottomMargin = getConfig().flowchart?.subGraphTitleMargin?.bottom || 0;
```
For consistency, we always use `??` instead of `||` for default fallbacks as `||` could lead to bugs in certain cases.

```suggestion
  const subGraphTitleTopMargin = getConfig().flowchart?.subGraphTitleMargin?.top ?? 0;
  const subGraphTitleBottomMargin = getConfig().flowchart?.subGraphTitleMargin?.bottom ?? 0;
```

sidharthv96 replied:
Can you also remove the duplicate call to getConfig()?

```suggestion
  const { flowchart } = getConfig();
  const subGraphTitleTopMargin = flowchart?.subGraphTitleMargin?.top ?? 0;
  const subGraphTitleBottomMargin = flowchart?.subGraphTitleMargin?.bottom ?? 0;
```

Or even better, make config a parameter to the function, the config is already available from your callsites, so we can pass it in here, eliminating the call altogether.

mathbraga replied:
I added the parameter as a destructered `{ flowchart }` that will come if passing `getConfig()` as argument when calling the function.

sidharthv96 commented on packages/mermaid/src/dagre-wrapper/index.js:
```diff
@@ -13,6 +13,9 @@ import { insertNode, positionNode, clear as clearNodes, setNodeElem } from './no
 import { log } from '../logger.js';
+import { getSubGraphTitleMargins } from '../utils/getSubGraphTitleMargins.js';
+
+const { subGraphTitleTotalMargin } = getSubGraphTitleMargins();
```
This should be moved into the function, otherwise it'll cause bugs when multiple diagrams are rendered.

mathbraga replied:
The reason I put this here was because the recursive calls would keep calling `getSubGraphTitleMargins()` unnecessarily. But if it causes bugs, then I think there's no option, I'll put it back in.

sidharthv96 replied:
If you add the config as parameter, then that call shouldn't be a big issue.

mathbraga replied:
Done.

sidharthv96 commented on packages/mermaid/src/utils.spec.ts:
```diff
@@ -594,3 +596,23 @@ describe('calculatePoint', () => {
     );
   });
 });
+
+describe('getSubGraphTitleMargins', () => {
+  it('should get subgraph title margins after config has been set', () => {
+    const config_0 = {
+      flowchart: {
+        subGraphTitleMargin: {
+          top: 10,
+          bottom: 5,
+        },
+      },
+    };
+
+    configApi.setSiteConfig(config_0);
+    expect(getSubGraphTitleMargins()).toEqual({
+      subGraphTitleTopMargin: 10,
+      subGraphTitleBottomMargin: 5,
+      subGraphTitleTotalMargin: 15,
+    });
+  });
+});
```
Would prefer this in `utils/subGraphTitleMargins.spec.ts` instead.
`utils.spec.ts` is overcrowded now, and there's a plan to split it up. 

Also, please rename the file `getSubGraphTitleMargins.ts` => `subGraphTitleMargins.ts`

mathbraga replied:
Done.

sidharthv96 commented on packages/mermaid/src/dagre-wrapper/clusters.js:
```diff
@@ -63,17 +64,22 @@ const rect = (parent, node) => {
       'transform',
       // This puts the labal on top of the box instead of inside it
-      'translate(' + (node.x - bbox.width / 2) + ', ' + (node.y - node.height / 2) + ')'
+      'translate(' +
+        (node.x - bbox.width / 2) +
+        ', ' +
+        (node.y - node.height / 2 + subGraphTitleTopMargin) +
+        ')'
```
```suggestion
      `translate(${node.x - bbox.width / 2}, ${node.y - node.height / 2 + subGraphTitleTopMargin})`
```

mathbraga replied:
Done.

sidharthv96 commented on packages/mermaid/src/dagre-wrapper/clusters.js:
```diff
@@ -63,17 +64,22 @@ const rect = (parent, node) => {
       'transform',
       // This puts the labal on top of the box instead of inside it
-      'translate(' + node.x + ', ' + (node.y - node.height / 2) + ')'
+      'translate(' + node.x + ', ' + (node.y - node.height / 2 + subGraphTitleTopMargin) + ')'
```
String templates here too.

mathbraga replied:
Done.

sidharthv96 reviewed (changes requested):
Nice work @mathbraga ! Love the tests.

Most of the comments are minor, but we need to dig in why it's not working for basic subgraphs _without_ external connections. 

[Live Editor](https://deploy-preview-5041--mermaid-js.netlify.app/edit#pako:eNpVT7tuwzAM_BWBcwy0QxZtTQpk8pSMWhiJtoVYpKFQKIIg_x45j7q96Xh3OJJX8BIILDRN49gLd7G3jk1FN8qPHzDra55xLsddxmk4RB2pxdxH_uPOUJms-fz4Lx5FVZI1a8ePPb_V5rB5JmtxPxeb9tJKKCM95RAzeY3Cj-BS-rXQzUK37whxcAwrSJQTxlDfu86GAx0okQNbacB8cuD4VnNYVPYX9mA1F1pBmQIqfUesNyWwHY5nut0Bca1ZqQ)

<img width="1512" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/8ee028ee-6827-4fa8-95ac-31a38a9594ef">

mathbraga commented:
@sidharthv96 First time I clicked your [link](https://deploy-preview-5041--mermaid-js.netlify.app/edit#pako:eNpVUD2vwjAM_CuRZyrB8JZswJOYOr2OWUJi2ojGroIjhBD_nbQFyrvpfD6dP-7g2CNoqKrKkGM6hVYbUgWnnq-us0le9YhLPh6SHbomSI-1TW2gr-4I4UGrzfq_eGQRjlr9GJrmfKJVs5udJbgdg1V9q9nnHmfZh4ROAtNkXEK3C90tdP-2IHlDsIKIKdrgy3n3sWFAOoxoQBfqbTobMPQoPpuF_27kQEvKuII8eCv4G2zZKb5F9EE41fO7pq89nlcuX_M) to the live preview I got the same result as shown on your image. Then I checked a second time and now it seems to work all of a sudden.

![image](https://github.com/mermaid-js/mermaid/assets/31048764/fb66abc4-86ad-4f2e-97cd-2db924ec7087)

sidharthv96 commented:
I noticed that setting the `subGraphTitleMargin.bottom` extends the bottom of the whole subgraph as well, that should be fixed. 

<img width="883" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/03b4bc97-ae99-43e0-b985-b1710252cf6e">

mathbraga commented:
@sidharthv96 I'm not sure I understand what you mean. If you set either top or bottom, both will extend the height of the whole subgraph. I implemented this way following the idea on the PR's design decision, where I stretch the subgraph and keep its children centered.

Should I discard the idea to keep the children centered? I did this just to maintain the layout consistent to the original without margins, and because of the simplicity of implementation, since I don't have to move any of the children nodes, edges, edge labels, etc.

Without keeping content centered, it looks something like this (top: 20, bottom:10):
![image](https://github.com/mermaid-js/mermaid/assets/31048764/16273b0d-857c-443b-b613-81a05154f165)

Is this what you mean?

sidharthv96 commented:
I think the extra space at the bottom came due to the workaround used in the issue. 

If we're adding a margin to the top/bottom of the title, I don't see the need to add extra padding to the bottom of the subgraph itself. 

In case we need padding for the subgraph itself, they should be exposed as extra options. 

Is keeping the children centred a requirement I missed?

Anyways, this is my personal opinion, we can discuss and finalize.

mathbraga commented:
Keeping content centered is not a requirement. I decided to do so just for consistency with the original layout.

So if the original looks centered, so will the layout with margins:
![image](https://github.com/mermaid-js/mermaid/assets/31048764/341b887d-5555-4eb1-9f55-2ba611262d9b)

But I think I get your point, so just to make sure we are on the same page, from the 2 layouts below, you prefer the one on the left? (both have same margin values)
![image](https://github.com/mermaid-js/mermaid/assets/31048764/ea9b0505-28be-4ca6-8f54-df509186d445)

mathbraga commented on packages/mermaid/src/dagre-wrapper/clusters.js:
```diff
@@ -63,17 +64,18 @@ const rect = (parent, node) => {
     .attr('width', width)
     .attr('height', node.height + padding);
 
+  const { subGraphTitleTopMargin } = getSubGraphTitleMargins(getConfig());
```
Done, I named the const `siteConfig`, is that an okay name? I avoided something generic like `config` since it already exists in the js default library.

sidharthv96 commented:
@mathbraga, yes I think the left one is the more accurate implementation of "_Subgraph Title_ margin", the bottom margin should be a separate config.

![image](https://github.com/mermaid-js/mermaid/assets/10703445/87a6aa74-46f4-403a-8cec-d63356cfd02a)

sidharthv96 commented on packages/mermaid/src/dagre-wrapper/index.js:
```diff
@@ -101,6 +103,7 @@ const recursiveRender = async (_elem, graph, diagramtype, id, parentCluster) =>
   log.info('Graph after layout:', graphlibJson.write(graph));
   // Move the nodes to the correct place
   let diff = 0;
+  const { subGraphTitleTotalMargin } = getSubGraphTitleMargins(getConfig());
```
Can you move this getConfig out of this recursive function? 
You can make it a parameter of recursiveRender and pass it from render function in this file, or add it as a parameter of the render function and pass the config from it's 3 callsites, where it is already available.

mathbraga replied:
Done, went with your first suggestion.

mathbraga commented:
@sidharthv96 It shouldn't have the excessive spacing now. I also added one more test for subgraphs with edge labels and added string templates on a few lines I spotted on a function I was working on. Tell me what you think.

aloisklink reviewed (approved):
Looks great to me! Thanks for adding this feature!

> I didn't add anything to the docs since I couldn't find how exactly it works with config changes. I looked around in the code and assumed config docs are generated automatically based on types.

Yep, that's right, they're generated automatically from the schema, so what you've done is perfect! This is what the docs will look like:

<details>
  <summary>Screenshot of auto-generated docs for flowchart.subGraphTitleMargin config</summary>

![image](https://github.com/mermaid-js/mermaid/assets/19716675/a0b2b706-04fb-4b23-ab86-d62e7083e3be)

</details>

sidharthv96 commented:
@mathbraga I'm afraid this PR has caused a regression. Do you think there is an easy fix for this?

https://github.com/mermaid-js/mermaid/pull/5188#issuecomment-1889781800

mathbraga commented:
@sidharthv96 The issue with the state diagram is a missplaced parenthesis at [line 193](https://github.com/mathbraga/mermaid/blob/5718be50ee9e10c38e84688da5ee692424ca24fa/packages/mermaid/src/dagre-wrapper/clusters.js#L193).

The issue with the sequence diagram are the changes made to [sequenceRenderer.ts](https://github.com/mermaid-js/mermaid/commit/42ac630304500e00a2b67395b79e03e13b57eec0#diff-1f8b0d46157ff209384aa6d146c2542142552e50ed6257ab525b1b789f5a0cde) that came from merging from develop at [42ac630](https://github.com/mermaid-js/mermaid/pull/5041/commits/42ac630304500e00a2b67395b79e03e13b57eec0).

Do I have to open a new PR?

sidharthv96 commented:
Can you see the Revert button in https://github.com/mermaid-js/mermaid/pull/5197 ?
If so, please use it and push the fixes to that PR.
